### PR TITLE
Prometheus backend

### PR DESCRIPTION
**Pull request by mmatloka**
https://github.com/softwaremill/sttp/issues/65

### CODE REVIEW

File: metrics/prometheus-backend/src/main/scala/com/softwaremill/sttp/prometheus/PrometheusBackend.scala [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  private[this] def createNewHistogram(name: String): Histogram = Histogram.build().name(name).help(name).register()
+}
+
+object PrometheusBackend {
+
+  val DefaultHistogramName = "sttp_request_latency"
+
+  def apply[R[_], S](delegate: SttpBackend[R, S],
+                     requestToHistogramNameMapper: Option[Request[_, S] => String] = None): SttpBackend[R, S] = {
+    // redirects should be handled before brave tracing, hence adding the follow-redirects backend on top
```
Comment (adamw) [inline metrics/prometheus-backend/src/main/scala/com/softwaremill/sttp/prometheus/PrometheusBackend.scala]: brave => prometheus?

File: docs/backends/prometheus.rst [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+This backend depends on `Prometheus JVM Client <https://github.com/prometheus/client_java>`_. Keep in mind this backend registers histograms and gathers request times, but you have to expose those metrics to `Prometheus <https://prometheus.io/>`_ e.g. using  `prometheus-akka-http <https://github.com/lonelyplanet/prometheus-akka-http>`_.
+
+The Prometheus backend wraps any other backend, for example::
+
+  implicit val sttpBackend = PrometheusBackend(AkkaHttpBackend())
+
+It uses by default ``sttp_request_latency`` histogram name, defined in ``PrometheusBackend.DefaultHistogramName``. It is possible to define custom histograms name by passing function mapping request to histogram name::
+
+  implicit val sttpBackend = PrometheusBackend(AkkaHttpBackend(), Some(request => request.uri.toString))
```
Comment (adamw) [inline docs/backends/prometheus.rst]: maybe a function `Request => Option[String]` would make more sense, allowing to measure the histogram only for some requests. 

Also, maybe a better example would be `request => request.uri.host`?

File: metrics/prometheus-backend/src/main/scala/com/softwaremill/sttp/prometheus/PrometheusBackend.scala [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+class PrometheusBackend[R[_], S] private (delegate: SttpBackend[R, S],
+                                          requestToHistogramNameMapper: Option[Request[_, S] => String])
+    extends SttpBackend[R, S] {
+
+  import PrometheusBackend._
+
+  private[this] val histograms: mutable.Map[String, Histogram] = new ConcurrentHashMap[String, Histogram]().asScala
+
+  override def send[T](request: Request[T, S]): R[Response[T]] = {
+    val histogramName = getHistogramName(request)
```
Comment (adamw) [inline metrics/prometheus-backend/src/main/scala/com/softwaremill/sttp/prometheus/PrometheusBackend.scala]: maybe also add a request counter and a requests-in-progress counter?

Comment (mmatloka) [reply metrics/prometheus-backend/src/main/scala/com/softwaremill/sttp/prometheus/PrometheusBackend.scala]: registering single histogram gives: `sttp_request_latency_sum`, `sttp_request_latency_count`, `sttp_request_latency `, `sttp_request_latency_bucket`, so we have a counter :)

regarding in-progress, hm, not so simple, we would need two counters (increment only) for 1) started and 2) finished https://prometheus.io/docs/concepts/metric_types/

Comment (adamw) [reply metrics/prometheus-backend/src/main/scala/com/softwaremill/sttp/prometheus/PrometheusBackend.scala]: a gauge? :)

File: docs/backends/prometheus.rst (hunk 2 above)
Comment (mmatloka) [reply docs/backends/prometheus.rst]: done

File: metrics/prometheus-backend/src/main/scala/com/softwaremill/sttp/prometheus/PrometheusBackend.scala (hunk 1 above)
Comment (mmatloka) [reply metrics/prometheus-backend/src/main/scala/com/softwaremill/sttp/prometheus/PrometheusBackend.scala]: done

File: metrics/prometheus-backend/src/main/scala/com/softwaremill/sttp/prometheus/PrometheusBackend.scala (hunk 3 above)
Comment (mmatloka) [reply metrics/prometheus-backend/src/main/scala/com/softwaremill/sttp/prometheus/PrometheusBackend.scala]: done

File: docs/backends/prometheus.rst [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+  implicit val sttpBackend = PrometheusBackend(AkkaHttpBackend(), request => Some(request.uri.host))
+
+You can disable request histograms by passing ``None`` returning function::
+
+  implicit val sttpBackend = PrometheusBackend(AkkaHttpBackend(), _ => None)
+
+This backend also offers ``Gauge`` with currently in-progress requests number. It uses by default ``sttp_requests_in_progress`` name, defined in ``PrometheusBackend.DefaultRequestsInProgressGaugeName``. It is possible to define custom gauge name by passing function mapping request to gauge name::
+
+   implicit val sttpBackend = PrometheusBackend(AkkaHttpBackend(), requestToHistogramNameMapper = request => Some(request.uri.host))
```
Comment (adamw) [inline docs/backends/prometheus.rst]: should be `requestToInProgressGaugeNameMapper`? :)

Comment (mmatloka) [reply docs/backends/prometheus.rst]: done

**Pull request merged**